include/proxy-wasm/context.h

```diff
@@ -0,0 +1,731 @@
+   */
+  virtual void onHttpCallResponse(uint32_t /* token */, uint32_t /* headers */,
+                                  uint32_t /* body_size */, uint32_t /* trailers */) {}
+
+  /**
+   * Called on a Root Context to indicate that an Inter-VM shared queue message has arrived.
+   * @token is the token returned by registerSharedQueue().
+   */
+  virtual void onQueueReady(uint32_t /* token */) { unimplemented(); }
+
+  // Context
+
+  /**
+   * Call when a stream has completed (both sides have closed) or on a Root Context when the VM is
+   * shutting down.
+   * @return true for stream contexts or for Root Context(s) if the VM can shutdown, false for Root
+   * Context(s) if the VM should wait until the Root Context calls the proxy_done() ABI call.  Note:
+   * the VM may (optionally) shutdown after some configured timeout even if the Root Context does
+   * not call proxy_done().
+   */
+  virtual bool onDone();
+
+  /**
+   * Call when a stream should log the final status of the stream. Occurs after onDone().
+   */
+  virtual void onLog();
+
+  /**
+   * Call when no further stream calls will occur.  This will cause the corresponding Context in the
+   * VM to be deleted.
+   */
+  virtual void onDelete();
+
+  /**
+   * Will be called on sever Wasm errors. Callees may report and handle the error (e.g. via an
+   * Exception) to prevent the proxy from crashing.
+   */
+  virtual void error(string_view message) {
+    std::cerr << message << "\n";
+    abort();
+  }
+
+  /**
+   * Called by all functions which are not overriden with a proxy-specific implementation.
+   */
+  virtual void unimplemented() { error("unimplemented proxy-wasm API"); }
+
+  //
+  // Calls from the VM>
+  // Note: most of these return a WasmResult, assume:
+  // @return a WasmResult with the status of the call.
+  //
+  //
+
+  /**
+   * Log a message.  Note: log_prefix() is *not* auto-prepended end may be useful to provide some
+   * context.
+   */
+  virtual WasmResult log(uint64_t /* level */, string_view /* message */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Enables a periodic timer with the given period or sets the period of an existing timer. Note:
+   * the timer is associated with the Root Context of whatever Context this call was made on and
+   * there is only one timer available per Root Context.
+   * @param period is the period of the periodic timer in milliseconds.
+   */
+  virtual WasmResult setTimerPeriod(std::chrono::milliseconds /* period */) {
+    unimplemented();
+    return WasmResult::Unimplemented;
+  }
+
+  /**
+   * Provides the current time in nanoseconds.
+   */
+  virtual uint64_t getCurrentTimeNanoseconds() {
+    struct timespec tpe;
+    clock_gettime(CLOCK_REALTIME, &tpe);
+    uint64_t t = tpe.tv_sec;
+    t *= 1000000000;
+    t += tpe.tv_nsec;
+    return t;
+  }
+
+  /**
+   * Provides the status of the call into the VM or out of the VM.
```

Comment:
What does this mean? Is this like errno/GetLastError()? Can you clarify?

Reply:
Yes, just like it.  Comments added.